File: src-tauri/src/commands/files.rs
use crate::error::Error;
use crate::state::AppState;
use crate::storage;
use std::path::PathBuf;
use std::process::Command;
use std::sync::Arc;
use tauri::State;

fn resolve_working_dir(group_dir: &str, project_cwd: &Option<String>) -> String {
    match project_cwd {
        Some(cwd) if !cwd.is_empty() => {
            let path = PathBuf::from(cwd);
            if path.is_absolute() {
                cwd.clone()
            } else {
                PathBuf::from(group_dir)
                    .join(cwd)
                    .to_string_lossy()
                    .to_string()
            }
        }
        _ => group_dir.to_string(),
    }
}

#[tauri::command]
pub fn resolve_project_working_dir(
    state: State<'_, Arc<AppState>>,
    group_id: String,
    project_id: String,
) -> Result<String, Error> {
    let config = state.config.lock().unwrap();
    let group = config
        .groups
        .iter()
        .find(|g| g.id == group_id)
        .ok_or_else(|| Error::GroupNotFound(group_id.clone()))?;
    let project = group
        .projects
        .iter()
        .find(|p| p.id == project_id)
        .ok_or_else(|| Error::ProjectNotFound(project_id.clone()))?;
    Ok(resolve_working_dir(&group.directory, &project.cwd))
}

#[tauri::command]
pub fn resolve_working_dir_by_project(
    state: State<'_, Arc<AppState>>,
    project_id: String,
) -> Result<String, Error> {
    let config = state.config.lock().unwrap();
    for group in &config.groups {
        if let Some(project) = group.projects.iter().find(|p| p.id == project_id) {
            return Ok(resolve_working_dir(&group.directory, &project.cwd));
        }
    }
    Err(Error::ProjectNotFound(project_id))
}

#[tauri::command]
pub fn open_file_in_editor(
    app_handle: tauri::AppHandle,
    file_path: String,
    line: Option<u32>,
    column: Option<u32>,
    working_dir: String,
) -> Result<(), Error> {
    // Resolve relative paths against working_dir
    let resolved = if PathBuf::from(&file_path).is_absolute() {
        PathBuf::from(&file_path)
    } else {
        PathBuf::from(&working_dir).join(&file_path)
    };

    // Canonicalize to resolve ../ and similar
    let resolved = resolved
        .canonicalize()
        .map_err(|_| Error::FileNotFound(file_path.clone()))?;

    let is_directory = resolved.is_dir();

    if !resolved.exists() {
        return Err(Error::FileNotFound(file_path));
    }

    // For directories, just open in file manager
    if is_directory {
        #[cfg(target_os = "linux")]
        {
            Command::new("xdg-open")
                .arg(resolved.to_string_lossy().to_string())
                .spawn()
                .map_err(|e| Error::SpawnError(format!("Failed to open directory: {}", e)))?;
        }
        #[cfg(target_os = "macos")]
        {
            Command::new("open")
                .arg(resolved.to_string_lossy().to_string())
                .spawn()
                .map_err(|e| Error::SpawnError(format!("Failed to open directory: {}", e)))?;
        }
        #[cfg(target_os = "windows")]
        {
            Command::new("explorer")
                .arg(resolved.to_string_lossy().to_string())
                .spawn()
                .map_err(|e| Error::SpawnError(format!("Failed to open directory: {}", e)))?;
        }
        return Ok(());
    }

    let resolved_str = resolved.to_string_lossy().to_string();

    // Get configured editor or fall back to environment
    let settings = storage::load_settings(&app_handle).unwrap_or_default();
    let editor = settings
        .editor
        .filter(|e| !e.is_empty())
        .or_else(|| std::env::var("VISUAL").ok())
        .or_else(|| std::env::var("EDITOR").ok())
        .unwrap_or_default();

    let editor_name = PathBuf::from(&editor)
        .file_name()
        .map(|n| n.to_string_lossy().to_string())
        .unwrap_or_default();

    let line_num = line.unwrap_or(1);
    let col_num = column.unwrap_or(1);

    let result = match editor_name.as_str() {
        "code" | "code-insiders" | "cursor" => Command::new(&editor)
            .arg("--goto")
            .arg(format!("{}:{}:{}", resolved_str, line_num, col_num))
            .spawn(),
        "zed" => Command::new(&editor)
            .arg(format!("{}:{}:{}", resolved_str, line_num, col_num))
            .spawn(),
        "subl" | "sublime_text" => Command::new(&editor)
            .arg(format!("{}:{}:{}", resolved_str, line_num, col_num))
            .spawn(),
        "vim" | "nvim" | "vi" => Command::new(&editor)
            .arg(format!("+{}", line_num))
            .arg(&resolved_str)
            .spawn(),
        "idea" | "goland" | "webstorm" | "phpstorm" | "pycharm" | "rustrover" | "clion" => {
            Command::new(&editor)
                .arg("--line")
                .arg(line_num.to_string())
                .arg("--column")
                .arg(col_num.to_string())
                .arg(&resolved_str)
                .spawn()
        }
        "emacs" | "emacsclient" => Command::new(&editor)
            .arg(format!("+{}:{}", line_num, col_num))
            .arg(&resolved_str)
            .spawn(),
        _ if !editor.is_empty() => Command::new(&editor).arg(&resolved_str).spawn(),
        _ => {
            // Fallback to platform default
            #[cfg(target_os = "linux")]
            {
                Command::new("xdg-open").arg(&resolved_str).spawn()
            }
            #[cfg(target_os = "macos")]
            {
                Command::new("open").arg(&resolved_str).spawn()
            }
            #[cfg(target_os = "windows")]
            {
                Command::new("cmd")
                    .args(["/C", "start", "", &resolved_str])
                    .spawn()
            }
        }
    };

    result.map_err(|e| Error::SpawnError(format!("Failed to open editor: {}", e)))?;

    Ok(())
}

#[tauri::command]
pub fn open_path(path: String) -> Result<(), Error> {
    let resolved = PathBuf::from(&path);

    if !resolved.exists() {
        return Err(Error::FileNotFound(path));
    }

    #[cfg(target_os = "linux")]
    {
        Command::new("xdg-open")
            .arg(&path)
            .spawn()
            .map_err(|e| Error::SpawnError(format!("Failed to open path: {}", e)))?;
    }
    #[cfg(target_os = "macos")]
    {
        Command::new("open")
            .arg(&path)
            .spawn()
            .map_err(|e| Error::SpawnError(format!("Failed to open path: {}", e)))?;
    }
    #[cfg(target_os = "windows")]
    {
        Command::new("explorer")
            .arg(&path)
            .spawn()
            .map_err(|e| Error::SpawnError(format!("Failed to open path: {}", e)))?;
    }

    Ok(())
}

#[tauri::command]
pub fn open_in_terminal(path: String) -> Result<(), Error> {
    let resolved = PathBuf::from(&path);

    if !resolved.exists() {
        return Err(Error::FileNotFound(path));
    }

    let resolved_str = resolved.to_string_lossy().to_string();

    #[cfg(target_os = "linux")]
    {
        // Try common Linux terminals
        let terminals = [
            ("kitty", vec!["--directory", &resolved_str]),
            ("alacritty", vec!["--working-directory", &resolved_str]),
            ("wezterm", vec!["start", "--cwd", &resolved_str]),
            ("gnome-terminal", vec!["--working-directory", &resolved_str]),
            ("konsole", vec!["--workdir", &resolved_str]),
            ("xfce4-terminal", vec!["--working-directory", &resolved_str]),
            ("mate-terminal", vec!["--working-directory", &resolved_str]),
            ("lxterminal", vec!["--working-directory", &resolved_str]),
            ("terminator", vec!["--working-directory", &resolved_str]),
            ("tilix", vec!["--working-directory", &resolved_str]),
            ("xterm", vec!["-e", "cd", &resolved_str, "&&", "bash"]),
        ];

        for (terminal, args) in &terminals {
            if which::which(terminal).is_ok() {
                let mut cmd = std::process::Command::new(terminal);
                cmd.args(args);
                match cmd.spawn() {
                    Ok(_) => return Ok(()),
                    Err(_) => continue,
                }
            }
        }

        return Err(Error::SpawnError(
            "No supported terminal found. Please install kitty, alacritty, wezterm, gnome-terminal, konsole, xfce4-terminal, or xterm.".to_string()
        ));
    }

    #[cfg(target_os = "macos")]
    {
        // Try common macOS terminals
        // First check for iTerm2
        let iterm_path = "/Applications/iTerm.app";
        if std::path::Path::new(iterm_path).exists() {
            std::process::Command::new("open")
                .args(&["-a", "iTerm", &resolved_str])
                .spawn()
                .map_err(|e| Error::SpawnError(format!("Failed to open iTerm: {}", e)))?;
            return Ok(());
        }

        // Check for WezTerm
        let wezterm_path = "/Applications/WezTerm.app";
        if std::path::Path::new(wezterm_path).exists() {
            std::process::Command::new("open")
                .args(&["-a", "WezTerm", &resolved_str])
                .spawn()
                .map_err(|e| Error::SpawnError(format!("Failed to open WezTerm: {}", e)))?;
            return Ok(());
        }

        // Check for kitty
        if which::which("kitty").is_ok() {
            std::process::Command::new("kitty")
                .args(&["--directory", &resolved_str])
                .spawn()
                .map_err(|e| Error::SpawnError(format!("Failed to open kitty: {}", e)))?;
            return Ok(());
        }

        // Check for alacritty
        if which::which("alacritty").is_ok() {
            std::process::Command::new("alacritty")
                .args(&["--working-directory", &resolved_str])
                .spawn()
                .map_err(|e| Error::SpawnError(format!("Failed to open alacritty: {}", e)))?;
            return Ok(());
        }

        // Fall back to Terminal.app
        std::process::Command::new("open")
            .args(&["-a", "Terminal", &resolved_str])
            .spawn()
            .map_err(|e| Error::SpawnError(format!("Failed to open Terminal: {}", e)))?;
        return Ok(());
    }

    #[cfg(target_os = "windows")]
    {
        // Try Windows Terminal first
        if which::which("wt").is_ok() {
            std::process::Command::new("wt")
                .args(&["-d", &resolved_str])
                .spawn()
                .map_err(|e| {
                    Error::SpawnError(format!("Failed to open Windows Terminal: {}", e))
                })?;
            return Ok(());
        }

        // Fall back to cmd.exe
        std::process::Command::new("cmd")
            .args(&["/C", "start", "cmd", "/K", "cd", "/d", &resolved_str])
            .spawn()
            .map_err(|e| Error::SpawnError(format!("Failed to open Command Prompt: {}", e)))?;
        return Ok(());
    }
}

File: src-electron/main/ipc/files.ts
/**
 * File-related IPC handlers
 * Ported from: src-tauri/src/commands/read_project_logs.rs, open_file_in_editor.rs, etc.
 */

import * as fs from 'fs';
import * as path from 'path';
import { spawn, execSync } from 'child_process';
import { ipcMain, shell } from 'electron';
import { IPC_CHANNELS } from '../../shared/events';
import { getState } from '../services/state';
import type { LogMessage } from '../../shared/types';

/**
 * Check if a command exists on the system
 */
function commandExists(command: string): boolean {
  try {
    const checkCmd =
      process.platform === 'win32' ? `where ${command}` : `which ${command}`;
    execSync(checkCmd, { stdio: 'ignore' });
    return true;
  } catch {
    return false;
  }
}

/**
 * Resolve the working directory for a project
 */
function resolveWorkingDir(groupDir: string, projectCwd: string | null): string {
  if (!projectCwd) {
    return groupDir;
  }

  // If projectCwd is absolute, use it directly
  if (path.isAbsolute(projectCwd)) {
    return projectCwd;
  }

  // Otherwise, resolve relative to group directory
  return path.resolve(groupDir, projectCwd);
}

export function registerFileHandlers(): void {
  // Read project logs from file
  ipcMain.handle(
    IPC_CHANNELS.READ_PROJECT_LOGS,
    async (
      _,
      args: {
        groupId: string;
        projectId: string;
      }
    ): Promise<LogMessage[]> => {
      const state = getState();

      // Try to get logs from the most recent session
      const logs = state.database.getRecentLogs(args.projectId, 1000);
      if (logs.length > 0) {
        return logs;
      }

      // Fall back to log file if exists
      const logFilePath = state.logFilePath(args.projectId);
      if (fs.existsSync(logFilePath)) {
        try {
          const content = fs.readFileSync(logFilePath, 'utf-8');
          return [
            {
              projectId: args.projectId,
              stream: 'stdout',
              data: content,
              timestamp: Date.now(),
            },
          ];
        } catch {
          // Ignore read errors
        }
      }

      return [];
    }
  );

  // Clear project logs
  ipcMain.handle(
    IPC_CHANNELS.CLEAR_PROJECT_LOGS,
    async (
      _,
      args: {
        groupId: string;
        projectId: string;
      }
    ): Promise<void> => {
      const state = getState();

      // Clear logs from database
      state.database.clearProjectLogs(args.projectId);

      // Clear log file if exists
      const logFilePath = state.logFilePath(args.projectId);
      if (fs.existsSync(logFilePath)) {
        fs.writeFileSync(logFilePath, '');
      }
    }
  );

  // Resolve project working directory
  ipcMain.handle(
    IPC_CHANNELS.RESOLVE_PROJECT_WORKING_DIR,
    async (
      _,
      args: {
        groupId: string;
        projectId: string;
      }
    ): Promise<string> => {
      const state = getState();
      const result = state.findProject(args.projectId);

      if (!result) {
        throw new Error(`Project not found: ${args.projectId}`);
      }

      return resolveWorkingDir(result.group.directory, result.project.cwd);
    }
  );

  // Resolve working directory by project ID
  ipcMain.handle(
    IPC_CHANNELS.RESOLVE_WORKING_DIR_BY_PROJECT,
    async (
      _,
      args: {
        projectId: string;
      }
    ): Promise<string> => {
      const state = getState();
      const result = state.findProject(args.projectId);

      if (!result) {
        throw new Error(`Project not found: ${args.projectId}`);
      }

      return resolveWorkingDir(result.group.directory, result.project.cwd);
    }
  );

  // Open file in editor
  ipcMain.handle(
    IPC_CHANNELS.OPEN_FILE_IN_EDITOR,
    async (
      _,
      args: {
        filePath: string;
        line?: number;
        column?: number;
        workingDir?: string;
      }
    ): Promise<void> => {
      const state = getState();
      const settings = state.database.getSettings();

      let editor = settings.editor;
      if (!editor) {
        // Try to detect system editor
        const envEditor = process.env.VISUAL || process.env.EDITOR;
        if (envEditor) {
          editor = envEditor;
        } else {
          // Check for common editors
          const commonEditors = [
            'code',
            'cursor',
            'zed',
            'windsurf',
            'nvim',
            'vim',
          ];
          for (const e of commonEditors) {
            if (commandExists(e)) {
              editor = e;
              break;
            }
          }
        }
      }

      if (!editor) {
        throw new Error(
          'No editor configured. Set an editor in settings or install VS Code, Cursor, or Zed.'
        );
      }

      // Build editor arguments
      const editorArgs: string[] = [];

      // Handle line/column for common editors
      if (args.line !== undefined) {
        const editorLower = editor.toLowerCase();
        if (
          editorLower.includes('code') ||
          editorLower.includes('cursor') ||
          editorLower.includes('windsurf')
        ) {
          // VS Code / Cursor format: --goto file:line:column
          const location = args.column
            ? `${args.filePath}:${args.line}:${args.column}`
            : `${args.filePath}:${args.line}`;
          editorArgs.push('--goto', location);
        } else if (editorLower.includes('zed')) {
          // Zed format: file:line:column
          const location = args.column
            ? `${args.filePath}:${args.line}:${args.column}`
            : `${args.filePath}:${args.line}`;
          editorArgs.push(location);
        } else if (
          editorLower.includes('vim') ||
          editorLower.includes('nvim')
        ) {
          // Vim format: +line file
          editorArgs.push(`+${args.line}`, args.filePath);
        } else {
          // Default: just open the file
          editorArgs.push(args.filePath);
        }
      } else {
        editorArgs.push(args.filePath);
      }

      // Spawn editor
      const cwd = args.workingDir || process.cwd();
      spawn(editor, editorArgs, {
        cwd,
        detached: true,
        stdio: 'ignore',
      }).unref();
    }
  );

  // Open path in system file manager
  ipcMain.handle(
    IPC_CHANNELS.OPEN_PATH,
    async (
      _,
      args: {
        path: string;
      }
    ): Promise<void> => {
      await shell.openPath(args.path);
    }
  );

  // Open path in terminal
  ipcMain.handle(
    IPC_CHANNELS.OPEN_IN_TERMINAL,
    async (
      _,
      args: {
        path: string;
      }
    ): Promise<void> => {
      const resolvedPath = path.resolve(args.path);

      if (!fs.existsSync(resolvedPath)) {
        throw new Error(`Path not found: ${args.path}`);
      }

      if (process.platform === 'linux') {
        // Try common Linux terminals
        const terminals: [string, string[]][] = [
          ['kitty', ['--directory', resolvedPath]],
          ['alacritty', ['--working-directory', resolvedPath]],
          ['wezterm', ['start', '--cwd', resolvedPath]],
          ['gnome-terminal', ['--working-directory', resolvedPath]],
          ['konsole', ['--workdir', resolvedPath]],
          ['xfce4-terminal', ['--working-directory', resolvedPath]],
          ['mate-terminal', ['--working-directory', resolvedPath]],
          ['lxterminal', ['--working-directory', resolvedPath]],
          ['terminator', ['--working-directory', resolvedPath]],
          ['tilix', ['--working-directory', resolvedPath]],
        ];

        for (const [terminal, termArgs] of terminals) {
          if (commandExists(terminal)) {
            spawn(terminal, termArgs, {
              detached: true,
              stdio: 'ignore',
            }).unref();
            return;
          }
        }

        throw new Error(
          'No supported terminal found. Please install kitty, alacritty, wezterm, gnome-terminal, konsole, or xfce4-terminal.'
        );
      } else if (process.platform === 'darwin') {
        // macOS
        const iTermPath = '/Applications/iTerm.app';
        if (fs.existsSync(iTermPath)) {
          spawn('open', ['-a', 'iTerm', resolvedPath], {
            detached: true,
            stdio: 'ignore',
          }).unref();
          return;
        }

        const wezTermPath = '/Applications/WezTerm.app';
        if (fs.existsSync(wezTermPath)) {
          spawn('open', ['-a', 'WezTerm', resolvedPath], {
            detached: true,
            stdio: 'ignore',
          }).unref();
          return;
        }

        if (commandExists('kitty')) {
          spawn('kitty', ['--directory', resolvedPath], {
            detached: true,
            stdio: 'ignore',
          }).unref();
          return;
        }

        if (commandExists('alacritty')) {
          spawn('alacritty', ['--working-directory', resolvedPath], {
            detached: true,
            stdio: 'ignore',
          }).unref();
          return;
        }

        // Fall back to Terminal.app
        spawn('open', ['-a', 'Terminal', resolvedPath], {
          detached: true,
          stdio: 'ignore',
        }).unref();
      } else if (process.platform === 'win32') {
        // Windows
        if (commandExists('wt')) {
          spawn('wt', ['-d', resolvedPath], {
            detached: true,
            stdio: 'ignore',
          }).unref();
          return;
        }

        // Fall back to cmd.exe
        spawn('cmd', ['/C', 'start', 'cmd', '/K', 'cd', '/d', resolvedPath], {
          detached: true,
          stdio: 'ignore',
        }).unref();
      }
    }
  );
}
